puffin/src/scope_details.rs

```diff
@@ -0,0 +1,239 @@
+    #[inline]
+    pub fn fetch_by_id(&self, scope_id: &ScopeId) -> Option<&Arc<ScopeDetails>> {
+        self.0.scope_id_to_details.get(scope_id)
+    }
+
+    /// Fetches scope details by scope name.
+    #[inline]
+    pub fn fetch_by_name(&self, scope_name: &str) -> Option<&ScopeId> {
+        self.0.type_to_scope_id.get(scope_name)
+    }
+
+    /// Insert a scope into the collection.
+    /// This method asserts the scope id is set which only puffin should do.
+    /// Custom sinks might use this method to store new scope details received from puffin.
+    pub fn insert(&mut self, scope_details: Arc<ScopeDetails>) -> Arc<ScopeDetails> {
+        let scope_id = scope_details
+            .scope_id
+            .expect("`ScopeDetails` missing `ScopeId`");
+
+        self.0
+            .type_to_scope_id
+            .insert(scope_details.name().clone(), scope_id);
+        self.0
+            .scope_id_to_details
+            .entry(scope_id)
+            .or_insert(scope_details)
+            .clone()
+    }
+
+    /// Manually register scope details. After a scope is inserted it can be reported to puffin.
+    pub(crate) fn register_user_scopes(
+        &mut self,
+        scopes: &[ScopeDetails],
+    ) -> Vec<Arc<ScopeDetails>> {
+        let mut new_scopes = Vec::with_capacity(scopes.len());
+        for scope_detail in scopes {
+            let new_scope_id = fetch_add_scope_id();
+            let scope = self.insert(Arc::new(
+                (*scope_detail).clone().with_scope_id(new_scope_id),
+            ));
+            new_scopes.push(scope);
+        }
+        new_scopes
+    }
+
+    /// Fetches all registered scopes and their ids.
+    /// Useful for fetching scope id by it's scope name.
+    /// For profiler scopes and user scopes this is the manual provided name.
+    /// For function profiler scopes this is the function name.
+    #[inline]
+    pub fn scopes_by_name(&self) -> &HashMap<Cow<'static, str>, ScopeId> {
+        &self.0.type_to_scope_id
+    }
+
+    /// Fetches all registered scopes.
+    /// Useful for fetching scope details by a scope id.
+    #[inline]
+    pub fn scopes_by_id(&self) -> &HashMap<ScopeId, Arc<ScopeDetails>> {
+        &self.0.scope_id_to_details
+    }
+}
+
+// Scopes are identified by user-provided name while functions are identified by the function name.
+#[derive(Debug, Clone, PartialEq, Hash, PartialOrd, Ord, Eq)]
+#[cfg_attr(
+    feature = "serialization",
+    derive(serde::Serialize, serde::Deserialize)
+)]
+pub enum ScopeType {
+    /// The scope is a function profile scope generated by `puffin::profile_function!`.
+    Function,
+    /// The scope is a profile scope inside a function generated by `puffin::profile_scope!` or registered manually.
+    Scope,
+}
+
+impl ScopeType {
+    pub fn type_str(&self) -> &'static str {
+        match self {
+            ScopeType::Function => "function scope",
+            ScopeType::Scope => "scope",
```

Comment:
The ui currently reads "scope type: scope" which is just confusing

```suggestion
            ScopeType::Function => "function",
            ScopeType::Scope => "named",
```